### PR TITLE
MAINT: start applying ruff/flake8-simplify rules (SIM)

### PR DESCRIPTION
**Pull request by DimitriPapadopoulos**
Will help with https://github.com/numpy/numpy/issues/24994.

### CODE REVIEW

**Comment by charris** [comment]
Thanks @DimitriPapadopoulos .

I find Ruff's tendency to stick computations in a return statement annoying.

File: numpy/_core/numerictypes.py [hunk 1]
```diff
@@ -223,12 +223,10 @@ def issctype(rep):
...
     try:
         res = obj2sctype(rep)
-        if res and res != object_:
-            return True
-        return False
     except Exception:
         return False
-
+    else:
+        return res and res != object_
```
Comment (seberg) [inline numpy/_core/numerictypes.py:229]: Oddly, this is one was not correct, it should `return bool()` if anything.

Comment (charris) [reply numpy/_core/numerictypes.py:229]: Hmm, good point. I think there was another that replaced True/False, but by a comparison.

**Comment by DimitriPapadopoulos** [comment]
> I find Ruff's tendency to stick computations in a return statement annoying.

I think that was not an automatic fix. I fixed some of the warnings manually. I should have made a difference between automatic fixes and manual fixes.

File: numpy/_core/numerictypes.py (hunk 1 above)
Comment (charris) [reply numpy/_core/numerictypes.py:229]: Looks like `bool()` doesn't work, it converts the dtypes to `np.bool`, so the resulting return types are a mixture of 'bool' and 'np.bool', which aren't the same. Not sure if that is a bug or not :)

Comment (DimitriPapadopoulos) [reply numpy/_core/numerictypes.py:229]: I think this was a manual fix. Indeed, it should be:
```
        return bool(res) and res != object_
```

Comment (DimitriPapadopoulos) [reply numpy/_core/numerictypes.py:229]: The `SIM` rules are not enabled by default, by the way. These made sense to me, but they can be reverted.

Comment (charris) [reply numpy/_core/numerictypes.py:229]: Fix in #26883. I just used the original and added a test.

Comment (seberg) [reply numpy/_core/numerictypes.py:229]: I guess there is an import of `np.bool` or `*` import somewhere...  I do think the builtin bool would work fine.  Could also have been `res is not None and res != np.object_` I think.
But doesn't matter, reverting is fine for now.

**Pull request merged**